src/blob/index/bptree/serializer.rs

```diff
@@ -0,0 +1,269 @@
+use super::prelude::*;
+
+struct HeaderStage {
+    header: IndexHeader,
+    key_size: usize,
+    filter_buf: Vec<u8>,
+}
+
+struct LeavesStage {
+    leaf_size: usize,
+    leaves_buf: Vec<u8>,
+    leaves_offset: u64,
+    headers_size: usize,
+}
+
+struct TreeStage {
+    metadata: TreeMeta,
+    meta_buf: Vec<u8>,
+    tree_buf: Vec<u8>,
+}
+
+pub(super) struct Serializer<'a> {
+    headers_btree: &'a InMemoryIndex,
+    header_stage: Option<HeaderStage>,
+    leaves_stage: Option<LeavesStage>,
+    tree_stage: Option<TreeStage>,
+}
+
+impl<'a> Serializer<'a> {
```

Comment:
Looks fine, but there is one problem: sequence of stages are checked at runtime, while it's possible to check at compile time. To do that, consider making stages contain needed info (like headers_btree), and make separate implementations, in which there will be only valid order of method invocations. E.g. some stub (optional) will contain method `header_stage`, which will produce instance of `HeaderStage`. And implementation of `HeaderStage` will contain `leaves_stage` method, producing `LeavesStage` and so on. That way you don't need to check, whether previous stages were completed - you know they were.